### PR TITLE
Add posibillty to use composition on memberstype

### PR DESCRIPTION
**Pull request by rammi987**
### Prerequisites
1) Start up Umbraco
2) Create Member type with basic fields like, firstname and lastname
3) create member type for a privatePerson, compostion with basicMember
4) create member type for a busneis person, composition with basicMember, and add field for VAT

Now you have the posibilty to reduce some controller logic, by using the interface BasicMember in ModelsBuilder.
And if you want to add phoneNumber til all memberTypes, you can add it to BasicMember and, both private and busneis will inherits it.

### Description
I have in a couple of time meet that i´m need to create diffent member type. But all the times I need to copy and past the field for one to the secound member type. I take a while when you have 5+ member types. Along with that they all need to have a set of basic infomation, and they need to by in sync all member types.

Also the code i nasty :( My latest ex is.
```
  var authenticatedMember = await _memberManager.GetCurrentMemberAsync();
  if (authenticatedMember is not null)
  {
      model.Email = authenticatedMember.Email ?? string.Empty;

      var member = _memberManager.AsPublishedMember(authenticatedMember);
      if (member is PrivateMember privateMember)
      {
          model.BillingAddress.FirstName = privateMember.FirstName;
          model.BillingAddress.LastName = privateMember.LastName;

          model.BillingAddress.Address = privateMember.Street ?? string.Empty;
          model.BillingAddress.ZipCode = privateMember.PostalCode ?? string.Empty;
          model.BillingAddress.City = privateMember.City ?? string.Empty;
          model.BillingAddress.Country = privateMember.Country?.Id ?? Guid.Empty;
          model.BillingAddress.PhoneNumber = privateMember.PhoneNumber ?? string.Empty;
      }
      else if (member is EmployeeMember employeeMember)
      {
          model.BillingAddress.FirstName = employeeMember.FirstName;
          model.BillingAddress.LastName = employeeMember.LastName;

          model.BillingAddress.Address = employeeMember.Street ?? string.Empty;
          model.BillingAddress.ZipCode = employeeMember.PostalCode ?? string.Empty;
          model.BillingAddress.City = employeeMember.City ?? string.Empty;
          model.BillingAddress.Country = employeeMember.Country?.Id ?? Guid.Empty;
          model.BillingAddress.PhoneNumber = employeeMember.PhoneNumber ?? string.Empty;
      }
      else if (member is ErhvervMember erhvervMember)
      {
          model.BillingAddress.FirstName = erhvervMember.FirstName;
          model.BillingAddress.LastName = erhvervMember.LastName;

          model.BillingAddress.Address = erhvervMember.Street ?? string.Empty;
          model.BillingAddress.ZipCode = erhvervMember.PostalCode ?? string.Empty;
          model.BillingAddress.City = erhvervMember.City ?? string.Empty;
          model.BillingAddress.Country = erhvervMember.Country?.Id ?? Guid.Empty;
          model.BillingAddress.PhoneNumber = erhvervMember.PhoneNumber ?? string.Empty;

          model.BillingAddress.VAT = erhvervMember.VAT ?? string.Empty;
          model.BillingAddress.CompanyName= erhvervMember.CompanyName ?? string.Empty;
      }
  }
```
It´s a huge waste of code block, when think of the only diff is that the ErhvervMember` needs to fill out the VAT and companyName. Where I with this PR could do this instead.

```
  var authenticatedMember = await _memberManager.GetCurrentMemberAsync();
  if (authenticatedMember is not null)
  {
      model.Email = authenticatedMember.Email ?? string.Empty;

      var member = _memberManager.AsPublishedMember(authenticatedMember);
      if (member is IBasicMember basicMember)
      {
          model.BillingAddress.FirstName = basicMember.FirstName;
          model.BillingAddress.LastName = basicMember.LastName;
          model.BillingAddress.Address = basicMember.Street ?? string.Empty;
          model.BillingAddress.ZipCode = basicMember.PostalCode ?? string.Empty;
          model.BillingAddress.City = basicMember.City ?? string.Empty;
          model.BillingAddress.Country = basicMember.Country?.Id ?? Guid.Empty;
          model.BillingAddress.PhoneNumber = basicMember.PhoneNumber ?? string.Empty;
      }
      
      if (member is ErhvervMember erhvervMember)
      {
          model.BillingAddress.VAT = erhvervMember.VAT ?? string.Empty;
          model.BillingAddress.CompanyName= erhvervMember.CompanyName ?? string.Empty;
      }
  }
```

### CODE REVIEW

**Comment by mikecp** [comment]
Hi @rammi987 ,

This looks like a great new feature! One of us will have a look quickly 😉

Cheers!

**Comment by rammi987** [comment]
Hello @lauraneto,
Is there any news to this PR? I would love to see it in 11.4.0 to refactor a lot of my code about shared member properties :)

File: src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs [hunk 1]
```diff
@@ -150,17 +152,16 @@ public IActionResult DeleteById(int id)
     ///     be looked up via the db, they need to be passed in.
     /// </param>
     /// <returns></returns>
-    public IActionResult GetAvailableCompositeMemberTypes(
-        int contentTypeId,
-        [FromQuery] string[] filterContentTypes,
-        [FromQuery] string[] filterPropertyTypes)
+    [HttpPost]
+    [Authorize(Policy = AuthorizationPolicies.TreeAccessMediaTypes)]
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs:156]: You don't need to add this policy here, this entire controller is already marked with this policy: 

```
[PluginController(Constants.Web.Mvc.BackOfficeApiArea)]
[Authorize(Policy = AuthorizationPolicies.TreeAccessMemberTypes)]
[ParameterSwapControllerActionSelector(nameof(GetById), "id", typeof(int), typeof(Guid), typeof(Udi))]
public class MemberTypeController : ContentTypeControllerBase<IMemberType>
```

So adding the attribute again for the method is redundant.

File: src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs [hunk 2]
```diff
@@ -150,17 +152,16 @@ public IActionResult DeleteById(int id)
     ///     be looked up via the db, they need to be passed in.
     /// </param>
     /// <returns></returns>
-    public IActionResult GetAvailableCompositeMemberTypes(
-        int contentTypeId,
-        [FromQuery] string[] filterContentTypes,
-        [FromQuery] string[] filterPropertyTypes)
+    [HttpPost]
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs:155]: Changing the http method for a public method is a breaking change.

File: src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs [hunk 3]
```diff
@@ -150,17 +152,16 @@ public IActionResult DeleteById(int id)
...
+    [Authorize(Policy = AuthorizationPolicies.TreeAccessMediaTypes)]
+    public IActionResult GetAvailableCompositeMemberTypes(GetAvailableCompositionsFilter filter)
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs:157]: Updating the method signature of a public method is a breaking change.

File: src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs [hunk 4]
```diff
@@ -172,6 +173,23 @@ public IActionResult DeleteById(int id)
...
 
+    /// <summary>
+    ///     Returns where a particular composition has been used
+    ///     This has been wrapped in a dto instead of simple parameters to support having multiple parameters in post request
+    ///     body
+    /// </summary>
+    /// <param name="filter"></param>
+    /// <returns></returns>
+    [HttpPost]
+    [Authorize(Policy = AuthorizationPolicies.TreeAccessMediaTypes)]
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs:184]: As above, adding this attribute is redundant as the whole controller already adheres to this policy.

File: src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs (hunk 2 above)
Comment (nul800sebastiaan) [reply src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs:155]: Also: you've defined the http method as `Post` but the method name is prefixed with `Get` - this is inconsistent. Looking at the use case here, you are indeed just getting data back so a `Get` is appropriate here, you're not posting back processed data that needs persisting.

File: src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs [hunk 5]
```diff
@@ -172,6 +173,23 @@ public IActionResult DeleteById(int id)
...
     }
 
+    /// <summary>
+    ///     Returns where a particular composition has been used
+    ///     This has been wrapped in a dto instead of simple parameters to support having multiple parameters in post request
+    ///     body
+    /// </summary>
+    /// <param name="filter"></param>
+    /// <returns></returns>
+    [HttpPost]
```
Comment (nul800sebastiaan) [inline src/Umbraco.Web.BackOffice/Controllers/MemberTypeController.cs:183]: As above, this is a `Get` request, but you added a `Post` attribute to it, so we're now in an ambiguous state, this looks like a pure `Get` request to me, so no need to force it to be a  `Post`.

**Review by nul800sebastiaan — COMMENTED** [review]
Hi @rammi987! Sorry for the delay, that was caused by my team having questions about breaking changes and me being on holiday, so I couldn't get back to them in a timely manner. In general this change looks good! I have left some comments about breaking changes and http verbs.  

If you can revert `GetAvailableCompositeMemberTypes` to the way it was before and change `GetWhereCompositionIsUsedInContentTypes` to not have a `[HttpPost]` attribute we should be in a good place to review this further.

Unfortunately 11.4 is already out in RC and the next version will be 12.0.0. The other bad news is that 12.0 will be closed for contribution soon as well since it's a new major that gets a longer Release Candidate period. So I think this won't make it in until v12.1.

**Comment by nul800sebastiaan** [comment]
Hi again @rammi987 - I've taken the liberty of updating the code to fix the breaking changes. Then I was going to test it out but.. am I missing something? I can not see the "Compositions" button, I can see that on the Document Types but not on Member Types. So I can't proceed with testing here. 🤔

image
image

**Comment by nul800sebastiaan** [comment]
@rammi987 I would just run `npm i && npm run build` on the command line in the `Umbraco.Web.UI.Client` folder. Make sure to also pull down your branch again to get my changes locally on your machine!

**Comment by nul800sebastiaan** [comment]
I also rebased all PRs on v12, so it might be necessary to do `npm cache clean --force` and delete the `npm_modules` directory to make sure everything is running perfectly :)

**Comment by rammi987** [comment]
Hello @nul800sebastiaan 
I manage to get my VS back up :)

I figure out why the button was missing, and while I was testing, I got a reponsecode 415 from `GetWhereCompositionIsUsedInMemberTypes` so I made som slight changes - hopes it´s not a breaking change :D

There is only one problem for now, I need some help with.
1) Create new memberType
2) In the process of creating a new type, hit the Composition button choose your memberType
3) save and everything works

1) Edit exiting memberType to contain a composition
2) in the process of edit the member type, hit the composition button.
- here is the problem all other types are unpickable :(

The reason for that is, if you create a normal membertype on save the MembersCommects comes as properties pr automatic.
Due to validation here Umbraco.Extensions.ContentTypeServiceExtensions.GetAvailableCompositeContentTypes line 117 the current type can´t container properties with the ones you want to composition with. In this cases it´s all other membertypes due to MembersCommects.

**Comment by nul800sebastiaan** [comment]
Very interesting @rammi987!! I've had some discussions with the devs at HQ and it seems like we don't know why we're forcing a `umbracoMemberComments` property on each member type. This is likely some leftover of what seemed like a good idea at the time. 

The usages of that field in the source code of Umbraco are not making it any clearer, the field is not used for anything at all, it's just.. There. Forced on all the types. We should probably remove it completely but that's a quite drastic change.

So here's the compromise: I saw that the field is being added during the initial creation of the database https://github.com/umbraco/Umbraco-CMS/blob/921ee4bc62fbff828137eed267b230b7d6f99b5b/src/Umbraco.Infrastructure/Migrations/Install/DatabaseDataCreator.cs#L1739 - so on the out-of-the-box first member type, I've left that in place, but I removed it from the list of items that are being forced during the creation of new member types. That was the only property being created anyway, so the list is now empty. But it's a very simple change that allows us to move on, in the future we can remove all of the code related to `umbracoMemberComments`.

So.. GOOD NEWS! Member types can now be composed just like document/media types. It seems to work absolutely great. I will let the test run finish and if the tests are green, we're good to go 🤞

**Comment by rammi987** [comment]
@nul800sebastiaan Sounds amazing!! And in my almost 10 year umbraco experience I have never ever used the umbraco member comment field - so I think it sound as a great simple change to empty out the list. I see forward to hear from you 💪

**Comment by nul800sebastiaan** [comment]
Tests failed 😭

It was because some of the tests were specifically testing that the `umbracoMemberComments` gets added. Other tests failed because the tab / propertyTypeGroup that the `umbracoMemberComments` property would be on was expected to be there. If there is 0 properties on a tab then the tab does not exist.

So I deleted some tests that are now irrelevant and updated the ones testing for number of tabs. Crossing fingers once again!

**Comment by nul800sebastiaan** [comment]
And again they failed: `/opt/hostedtoolcache/dotnet/sdk/7.0.203/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.ApiCompat.ValidatePackage.targets(39,5): error CP0002: Member 'Umbraco.Cms.Tests.Integration.Umbraco.Infrastructure.Persistence.Repositories.MemberTypeRepositoryTest.Built_In_Member_Type_Properties_Are_Automatically_Added_When_Creating()' exists on [Baseline] lib/net7.0/Umbraco.Tests.Integration.dll but not on lib/net7.0/Umbraco.Tests.Integration.dll [/home/vsts/work/1/s/tests/Umbraco.Tests.Integration/Umbraco.Tests.Integration.csproj]`

It is marked as a breaking change as we ship the integration tests, because it contains the setup to allow you to integration test your own solutions. 
I reintroduced the tests and just removed their functionality completely, adding a link to this PR to clarify why they are not doing anything.

Crossing fingers AND toes for this new test run 😂

**Comment by ronaldbarendse** [comment]
Awesome work @rammi987!👍🏻 This can indeed really save a lot of repetition when setting up multiple member types and it's nice to align this feature with the content and media types 🥳

One thing I want to mention though, is that you can also [extend the Models Builder classes](https://docs.umbraco.com/umbraco-cms/reference/templating/modelsbuilder/understand-and-extend#extending) with your own custom interface to avoid the verbose code from your examples. If you have 2 member types (e.g. 'Private person' and 'Business person') that both have a 'First name' and 'Last name' property, you can create an interface with these properties and extend the generated model classes using:

```c#
namespace Umbraco.Cms.Web.Common.PublishedModels;

public interface IPersonMember
{
    string FirstName { get; }
    string LastName { get; }
}

public partial class PrivatePerson : IPersonMember
{ }

public partial class BusinessPerson : IPersonMember
{ }
```

And then use that interface in a similar way as using the automatically generated composition interface:
```c#
var authenticatedMember = await _memberManager.GetCurrentMemberAsync();
if (authenticatedMember is not null)
{
    model.Email = authenticatedMember.Email ?? string.Empty;

    var member = _memberManager.AsPublishedMember(authenticatedMember);
    if (member is IPersonMember personMember)
    {
        model.BillingAddress.FirstName = personMember.FirstName;
        model.BillingAddress.LastName = personMember.LastName;
    }
}
```

The nice thing is that these approaches can be used together as well! Using compositions will make it easier to apply changes to multiple member types and automatically generate the composition interface, but if you want to e.g. use different aliases, names, descriptions or even property editors in the different member types, you can create the interface and extend the models manually (and map the interface properties however you wish).

@AndyButland We probably need to update Deploy to support deploying member types using compositions (or at least check whether this still works) ⚠️

**Pull request merged**